python/src/spark_rapids_ml/tuning.py

```diff
@@ -0,0 +1,167 @@
+#
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from multiprocessing.pool import ThreadPool
+from typing import TYPE_CHECKING, Callable, List, Sequence, Tuple, cast
+
+import numpy as np
+from pyspark import inheritable_thread_target
+from pyspark.ml import Estimator, Model, Transformer
+from pyspark.ml.evaluation import Evaluator, MulticlassClassificationEvaluator
+from pyspark.ml.tuning import CrossValidator as SparkCrossValidator
+from pyspark.ml.tuning import CrossValidatorModel
+from pyspark.sql import DataFrame
+
+from spark_rapids_ml.classification import (
+    RandomForestClassificationModel,
+    RandomForestClassifier,
+)
+
+if TYPE_CHECKING:
+    from pyspark.ml._typing import ParamMap
+
+
+# Copied from pyspark.ml.tuning
+def _parallelFitTasks(
+    est: Estimator,
+    train: DataFrame,
+    eva: Evaluator,
+    validation: DataFrame,
+    epm: Sequence["ParamMap"],
+    collectSubModel: bool,
+) -> List[Callable[[], Tuple[int, float, Transformer]]]:
+    """
+    Creates a list of callables which can be called from different threads to fit and evaluate
+    an estimator in parallel. Each callable returns an `(index, metric)` pair.
+
+    Parameters
+    ----------
+    est : :py:class:`pyspark.ml.baseEstimator`
+        he estimator to be fit.
+    train : :py:class:`pyspark.sql.DataFrame`
+        DataFrame, training data set, used for fitting.
+    eva : :py:class:`pyspark.ml.evaluation.Evaluator`
+        used to compute `metric`
+    validation : :py:class:`pyspark.sql.DataFrame`
+        DataFrame, validation data set, used for evaluation.
+    epm : :py:class:`collections.abc.Sequence`
+        Sequence of ParamMap, params maps to be used during fitting & evaluation.
+    collectSubModel : bool
+        Whether to collect sub model.
+
+    Returns
+    -------
+    tuple
+        (int, float, subModel), an index into `epm` and the associated metric value.
+    """
+    modelIter = est.fitMultiple(train, epm)
```

Comment:
I wonder if we should enable the case where multiple spark-rapids-ml clusters each with num_workers GPUs are fitting in parallel on portions of the parameters?   This could be more performant if say the training data fits into fewer GPUs (e.g. 1 gpu) than are in the whole spark cluster.    I guess this would require changing fitMultiple.   Maybe future version.

Reply:
Yes, we can try this to see if it can help to performance.